File: packages/shared/src/state-series.ts
import {z} from 'zod';
export const stateSeriesSchema = z.object({version:z.literal(1),samples:z.array(z.object({at:z.string().datetime({offset:true}),durationMs:z.number().int().min(0).max(300000),idleSeconds:z.number().int().min(0).max(Number.MAX_SAFE_INTEGER).optional(),availableStorageBytes:z.number().int().min(0).max(Number.MAX_SAFE_INTEGER).optional()}).strict()).min(1).max(720)}).strict();
/** Exact state comparison only. Clocks identify observations, not state changes. */
export interface StateSample {at:string;durationMs:number;idleSeconds?:number;availableStorageBytes?:number}
export interface StateSeries {version:1;samples:StateSample[]}
type StateEvent={id:string;deviceId:string;capturedAt:string;durationMs:number;source:string;imageBase64?:string;imageMime?:string|null;ocrText?:string;stateSeries?:StateSeries;metadata?:any};
function canonical(value:any):string {if(Array.isArray(value))return '['+value.map(canonical).join(',')+']';if(value&&typeof value==='object')return '{'+Object.keys(value).filter(k=>value[k]!==undefined).sort().map(k=>JSON.stringify(k)+':'+canonical(value[k])).join(',')+'}';return JSON.stringify(value);}
export function stateOnly(event:StateEvent):boolean {return !event.imageBase64&&!event.imageMime&&!event.ocrText&& (event.source==='activity'||event.source==='screen'&&event.metadata?.capture?.deduplication?.duplicate===true);}
export function stateKey(event:StateEvent):string {
 const {id,capturedAt,durationMs,stateSeries,...rest}=structuredClone(event);
 if(rest.metadata){delete rest.metadata.observedAt;if(rest.metadata.state){delete rest.metadata.state.idleSeconds;delete rest.metadata.state.availableStorageBytes;}if(rest.metadata.observation)delete rest.metadata.observation.elapsedRealtimeMs;if(rest.metadata.media)delete rest.metadata.media.observedAt;}
 return canonical(rest);
}
function observation(event:StateEvent):StateSample {const s=event.metadata?.state;return {at:event.capturedAt,durationMs:event.durationMs,...(s?.idleSeconds===undefined?{}:{idleSeconds:s.idleSeconds}),...(s?.availableStorageBytes===undefined?{}:{availableStorageBytes:s.availableStorageBytes})};}
export function samples(event:StateEvent):StateSample[] {return event.stateSeries?.samples??[observation(event)];}
export function extendState<T extends StateEvent>(previous:T|undefined,next:T):T {
 if(!stateOnly(next))return next;
 const observations=previous?samples(previous):[];const last=observations.at(-1);const gap=last?Date.parse(next.capturedAt)-Date.parse(last.at):Infinity;
 if(previous&&stateOnly(previous)&&stateKey(previous)===stateKey(next)&&gap>0&&gap<=300000&&observations.length<720&&Date.parse(next.capturedAt)-Date.parse(previous.capturedAt)<=21600000){return {...previous,stateSeries:{version:1,samples:[...observations,observation(next)]}};}
 return {...next,stateSeries:{version:1,samples:[observation(next)]}};
}
export function isStateExtension(previous:StateEvent,next:StateEvent):boolean {
 if(!previous.stateSeries||!next.stateSeries||previous.id!==next.id||previous.capturedAt!==next.capturedAt||!stateOnly(next)||stateKey(previous)!==stateKey(next))return false;
 const a=samples(previous),b=samples(next);return b.length>=a.length&&a.every((v,i)=>canonical(v)===canonical(b[i]));
}

File: apps/android/app/src/main/java/dev/mote/collector/StateSeries.kt
package dev.mote.collector

import org.json.JSONObject
import org.json.JSONArray
import java.time.Instant

/** Exact structural state equality. No interpretation of contents or user intent. */
object StateSeries {
    private fun eligible(v: JSONObject) = !v.has("imageMime") && !v.has("imageBase64") && v.optString("ocrText").isEmpty() && (v.optString("source") == "activity" || v.optString("source") == "screen" && v.optJSONObject("metadata")?.optJSONObject("capture")?.optJSONObject("deduplication")?.optBoolean("duplicate") == true)
    private fun key(v: JSONObject): String {
        val copy = JSONObject(v.toString()); listOf("id", "capturedAt", "durationMs", "stateSeries").forEach(copy::remove)
        copy.optJSONObject("metadata")?.let { m -> m.remove("observedAt"); m.optJSONObject("state")?.let { it.remove("idleSeconds"); it.remove("availableStorageBytes") }; m.optJSONObject("observation")?.remove("elapsedRealtimeMs"); m.optJSONObject("media")?.remove("observedAt") }
        return SourceRules.canonical(copy)
    }
    fun extend(previous: JSONObject?, next: JSONObject): JSONObject {
        if (!eligible(next)) return next
        val samples = previous?.optJSONObject("stateSeries")?.optJSONArray("samples")
        val at = Instant.parse(next.getString("capturedAt")).toEpochMilli()
        val last = samples?.takeIf { it.length() > 0 }?.getJSONObject(samples.length() - 1)
        val gap = last?.let { at - Instant.parse(it.getString("at")).toEpochMilli() } ?: Long.MAX_VALUE
        val observation = JSONObject().put("at", next.getString("capturedAt")).put("durationMs", next.optLong("durationMs"))
        next.optJSONObject("metadata")?.optJSONObject("state")?.let { state -> listOf("idleSeconds", "availableStorageBytes").forEach { if (state.has(it)) observation.put(it, state.get(it)) } }
        if (previous != null && samples != null && eligible(previous) && key(previous) == key(next) && gap in 1..300000 && samples.length() < 720 && at - Instant.parse(previous.getString("capturedAt")).toEpochMilli() <= 21600000) {
            return JSONObject(previous.toString()).apply { getJSONObject("stateSeries").getJSONArray("samples").put(observation) }
        }
        return JSONObject(next.toString()).put("stateSeries", JSONObject().put("version", 1).put("samples", JSONArray().put(observation)))
    }
}
